src/group_theory/iwasawa.lean

```diff
@@ -0,0 +1,487 @@
+   function.surjective φ → is_commutative G (*) → is_commutative H (*) :=
+begin
+  intros is_surj is_comm,
+  apply is_commutative.mk,
+  intros a b,
+    obtain ⟨a', ha'⟩ := is_surj a, obtain ⟨b', hb'⟩ := is_surj b,
+    rw ← ha', rw ← hb',
+    let z := ⇑φ, let z₂ := φ.to_fun, have : z = z₂  , by refl,
+    rw ← mul_hom.map_mul φ a' b',
+    rw ← mul_hom.map_mul φ b' a',
+    apply φ.congr_arg,
+    refine is_commutative.cases_on is_comm _, intro,
+    exact comm a' b',
+end
+
+/- No more used
+
+/-- f(f⁻¹(t)) = f.range ∩ t  -/
+lemma image_preimage_eq {α β : Type*} (f: α → β) (t: set β):
+  set.image f (set.preimage f t) = set.range f ∩ t :=
+begin
+    sorry,
+end
+
+lemma image_preimage_eq' {α β : Type*} (f: α → β) (t: set β):
+  set.image f (set.preimage f t) = set.range f ∩ t :=
+  by rw [← set.image_univ, ← set.image_inter_preimage f set.univ t, set.univ_inter]
+
+/-- If f is surjective, then f(f⁻¹(t)) = t -/
+--  is in mathlib :  set.image_preimage_eq
+lemma image_preimage_eq_of_surjective' {α β : Type*} (f : α → β) (t : set β) :
+  function.surjective f → set.image f (set.preimage f t) = t :=  λ hf,
+  by rw [image_preimage_eq f t, set.range_iff_surjective.2 hf,  set.univ_inter]
+ -/
+end Basic_Lemmas
+
+section Primitive_And_Transitive_Actions
+open_locale big_operators pointwise
+open function
+
+open set mul_action subgroup
+
+variables (G : Type*) [group G] (X : Type*) [mul_action G X]
+
+/-- Transitive actions -/
+structure is_transitive
+extends is_pretransitive G X : Prop :=
+(is_nonempty : nonempty X)
+
+/-
+lemma nonempty_of_transitive : is_transitive G X → nonempty X
+:= λ h, h.is_nonempty
+
+lemma pretransitive_of_transitive : is_transitive G X → is_pretransitive G X :=
+λ h, h.is_pretransitive
+-/
+
+variables (G X)
+structure is_primitive
+extends is_transitive G X : Prop :=
+(has_maximal_stabilizers: ∀ x:X, (stabilizer G x).is_maximal)
+
+/-
+lemma nonempty_of_primitive : is_transitive G X → nonempty X
+:= λ h, h.is_nonempty
+
+lemma transitive_of_primitive : is_primitive G X → is_transitive G X
+:= λ h, is_transitive.mk h.is_nonempty h.is_pretransitive
+
+lemma has_maximal_stabilizer : is_primitive G X →  ∀ (x : X), (stabilizer G x).is_maximal
+:= λ h, h.has_maximal_stabilizers
+-/
+
+/-- A 2-transitive action is primitive -/
+/- Part of the proof establishes that stabilizers of n-transitive
+actions are (n-1)-transitive. Todo: rewrite using this. -/
+theorem is_primitive_of_two_trans
+  : (∀ (x₁ x₂ y₁ y₂ : X) (hx: x₁ ≠ x₂) (hy: y₁ ≠ y₂), ∃ g: G,
+  g • x₁  = y₁ ∧ g • x₂ = y₂ ) →
+  nontrivial X → is_primitive G X :=
+begin
+    intros h2X hX,
+
+    have is_trans : is_transitive G X ,
+    { apply is_transitive.mk,
+      apply is_pretransitive.mk,
+      { intros x y,
+        /- on veut applique x sur y
+          mais il faut un élément x' ≠ x à appliquer sur un élément y' ≠ y    -/
+        obtain ⟨x', hx'⟩ := @exists_ne _ hX x,
+        obtain ⟨y', hy'⟩ := @exists_ne _ hX y,
```

Comment:
Does this work?
```suggestion
        obtain ⟨x', hx'⟩ := exists_ne x,
        obtain ⟨y', hy'⟩ := exists_ne y,
```

Reply:
No ! I get the following error message : 

```
failed to synthesize type class instance for
G : Type u_1,
_inst_1 : group G,
X : Type u_2,
_inst_2 : mul_action G X,
h2X :
  ∀ (x₁ x₂ y₁ y₂ : X),
    x₁ ≠ x₂ → y₁ ≠ y₂ → (∃ (g : G), g • x₁ = y₁ ∧ g • x₂ = y₂),
hX : nontrivial X,
x y : X
⊢ nontrivial X
```

which I interpret as Lean telling me it didn't find by itself that it could use `hX`.

Reply:
The issue seems to be that `hX` is not an instance. See my recent suggestion for a possible fix.